plugins/superio/fu-superio-device.c

```diff
@@ -0,0 +1,327 @@
+	return fu_superio_device_outb (self, self->cmd_port, data, error);
+}
+
+static gboolean
+fu_superio_device_ec_read (FuSuperioDevice *self, guint8 *data, GError **error)
+{
+	if (!fu_superio_device_wait_for (self, KB_OBF, TRUE, error))
+		return FALSE;
+	return fu_superio_device_inb (self, self->data_port, data, error);
+}
+
+static gboolean
+fu_superio_device_ec_write (FuSuperioDevice *self, guint8 data, GError **error)
+{
+	if (!fu_superio_device_wait_for (self, KB_IBF, FALSE, error))
+		return FALSE;
+	return fu_superio_device_outb (self, self->data_port, data, error);
+}
+
+static gboolean
+fu_superio_device_ec_flush (FuSuperioDevice *self, GError **error)
+{
+	guint8 status = 0x00;
+	g_autoptr(GTimer) timer = g_timer_new ();
+	do {
+		guint8 unused = 0;
+		if (!fu_superio_device_inb (self, self->cmd_port, &status, error))
+			return FALSE;
+		if ((status & KB_OBF) == 0)
+			break;
+		if (!fu_superio_device_inb (self, self->data_port, &unused, error))
+			return FALSE;
+		if (g_timer_elapsed (timer, NULL) > FU_PLUGIN_SUPERIO_TIMEOUT) {
+			g_set_error_literal (error,
+					     G_IO_ERROR,
+					     G_IO_ERROR_TIMED_OUT,
+					     "timed out whilst waiting for flush");
+			return FALSE;
+		}
+	} while (TRUE);
+	return TRUE;
+}
+
+static gboolean
+fu_superio_device_ec_get_param (FuSuperioDevice *self, guint8 param, guint8 *data, GError **error)
+{
+	if (!fu_superio_device_ec_cmd (self, 0x80, error))
+		return FALSE;
+	if (!fu_superio_device_ec_write (self, param, error))
+		return FALSE;
+	return fu_superio_device_ec_read (self, data, error);
+}
+
+#if 0
+static gboolean
+fu_superio_device_ec_set_param (FuSuperioDevice *self, guint8 param, guint8 data, GError **error)
+{
+	if (!fu_superio_device_ec_cmd (self, 0x81, error))
+		return FALSE;
+	if (!fu_superio_device_ec_write (self, param, error))
+		return FALSE;
+	return fu_superio_device_ec_write (self, data, error);
+}
+#endif
+
+static gchar *
+fu_superio_device_ec_get_str (FuSuperioDevice *self, guint8 idx, GError **error)
+{
+	GString *str = g_string_new (NULL);
+	if (!fu_superio_device_ec_cmd (self, idx, error))
+		return FALSE;
+	for (guint i = 0; i < 0xff; i++) {
+		guint8 c = 0;
+		if (!fu_superio_device_ec_read (self, &c, error))
+			return FALSE;
+		if (c == '$')
+			break;
+		g_string_append_c (str, c);
+	}
+	return g_string_free (str, FALSE);
+}
+
+static gboolean
+fu_superio_device_probe (FuDevice *device, GError **error)
+{
+	FuSuperioDevice *self = FU_SUPERIO_DEVICE (device);
+	guint8 size_tmp = 0;
+	g_autofree gchar *name = NULL;
+	g_autofree gchar *version = NULL;
+	g_autofree gchar *guid = NULL;
+
+	/* check ID is correct */
+	if (!fu_superio_device_check_id (self, error)) {
+		g_prefix_error (error, "failed to probe id: ");
+		return FALSE;
+	}
+
+	/* use the chipset name as the ID and for the GUID */
+	fu_device_set_id (FU_DEVICE (self), self->chipset);
+	guid = g_strdup_printf ("SuperIO-%s", self->chipset);
+	fu_device_add_guid (FU_DEVICE (self), guid);
```

Comment:
You already have an argument to this function of `device`, I think you can just use that rather than `FU_DEVICE(self)`

Reply:
Will fix.

Reply:
Fixed.